### PR TITLE
localstore: fix index leak when NN chunks enter localstore

### LINKED ISSUE #1920: problem with GC
As discussed in the #core chat room, it appears that there are still bugs with localstore GC, which result in the following storage indices (`swarm-private-38` - broken ; `swarm-private-0` - correct):

```
➜  ~ echo '{"jsonrpc":"2.0","method":"bzz_storageIndices","id":1}' | websocat ws://127.0.0.1:8001/api/v1/namespaces/pushsync/pods/http:swarm-private-38:8546/proxy/ -n --one-message | jq

{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "gcExcludeIndex": 0,
    "gcIndex": 188484,
    "gcSize": 188484,
    "pinIndex": 0,
    "pullIndex": 1036970,
    "pushIndex": 0,
    "retrievalAccessIndex": 188484,
    "retrievalDataIndex": 1073923
  }
}
➜  ~ echo '{"jsonrpc":"2.0","method":"bzz_storageIndices","id":1}' | websocat ws://127.0.0.1:8001/api/v1/namespaces/pushsync/pods/http:swarm-private-0:8546/proxy/ -n --one-message | jq

{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "gcExcludeIndex": 0,
    "gcIndex": 192540,
    "gcSize": 192540,
    "pinIndex": 0,
    "pullIndex": 192540,
    "pushIndex": 0,
    "retrievalAccessIndex": 192540,
    "retrievalDataIndex": 192540
  }
}
➜  ~
```

One idea by @janos is that this is due to `handleChunkDelivery` and namely the `ModePutRequest` at:

```
	if po >= depth || peerPO < po {
		mode = chunk.ModePutSync
	} else {
		// do not sync if peer that is sending us a chunk is closer to the chunk then we are
		mode = chunk.ModePutRequest
	}
```

### PR DESCRIPTION
**Pull request by acud**
This PR fixes an index leak that was manifesting when chunks that would enter the node's local storage would, under certain connectivity conditions be not added to the GC index.

This is due to the fact that the uploader node would be the closest node to the chunk, thus again, under certain connectivity condition this would result in the chunk never syncing, and the `retrievalDataIndex` eventually growing out of bound, since no matter how many chunks in lower bins would be garbage collected, the higher bins would always make the store size eventually overflow.

This fix introduces a pluggable function to localstore that is called when a chunk enters the local storage.

During implementation several weaknesses were unraveled:
1. a node with very little storage could GC the content uploaded before it is pull synced
2. as a result, anonymous content syncing on light nodes could be severely damaged
3. a flag indicating that the node is actively waiting for an upload to sync should probably be added to Swarm in order to short-circuit this mechanism for ephemeral nodes

fixes https://github.com/ethersphere/swarm/issues/1920

### CODE REVIEW

**Comment by janos** [comment]
The code cannot be compiled:

```
storage/localstore/mode_put_test.go:325:19: cannot use func literal (type func(chunk.Address) bool) as type func([]byte) bool in field value
```

File: sync_test.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
+package swarm
```
Comment (janos) [inline sync_test.go]: This file is probably committed by mistake?

I wrote this test only to manually debug the issue. It is not a test that does any validation and contains a lot of print function calls. It surely is not in the mergeable condition.

File: storage/localstore/mode_put_test.go [hunk 2]
```diff
@@ -318,6 +318,47 @@ func TestModePut_sameChunk(t *testing.T) {
 	}
 }
 
+// TestModePutSync_addToGc validates ModePut* with PutSetCheckFunc stub results
+// in the added chunk to show up in GC index
+func TestModePut_addToGc(t *testing.T) {
+	// PutSetCheckFunc will tell localstore to always Set the chunks that enter
+	opts := &Options{PutSetCheckFunc: func(_ []byte) bool { return true }}
+	for _, m := range []chunk.ModePut{
```
Comment (janos) [inline storage/localstore/mode_put_test.go]: This test should contain checks that chunk is not added to the GC if PutSetCheckFunc returns false.

It also should test chunks that are put to the database, but already accessed. It is important that both access and gc indexes are properly updated in this case.

File: network/kademlia.go [hunk 3]
```diff
@@ -758,6 +758,13 @@ func (k *Kademlia) IsClosestTo(addr []byte, filter func(*BzzPeer) bool) (closest
 	return closest
 }
 
+func (k *Kademlia) IsAddressWithinDepth(addr []byte) bool {
```
Comment (janos) [inline network/kademlia.go]: I would maybe call it just `IsWithinDepth` as it should be clear that the address is checked.

And maybe to add a comment to the exported function.

File: storage/localstore/localstore.go [hunk 4]
```diff
@@ -133,6 +135,10 @@ type Options struct {
 	// MetricsPrefix defines a prefix for metrics names.
 	MetricsPrefix string
 	Tags          *chunk.Tags
+	// PutSetCheckFunc is a function called after a Put of a chunk
+	// to verify whether that chunk needs to be Set and added to
+	// garbage collection index too
+	PutSetCheckFunc func([]byte) bool
```
Comment (janos) [inline storage/localstore/localstore.go]: This function has to do with adding a chunk to the GC, not to any Set method call. A name like `PutToGCCheck` may be more appropriate, or something similar?

**Review by janos — CHANGES_REQUESTED** [review]

**Review by janos — APPROVED** [review]
LGTM. 👍

**Review by zelig — APPROVED** [review]

**Review by nolash — COMMENTED** [review]
I see @zelig reviewed it while I had breakfast halfway. Anyway, curious about the negative bin id

Also curious about the description;

> as a result, anonymous content syncing on light nodes could be severely damaged

what is anonmous content syncing on light nodes?!

> under certain connectivity condition

I would like descriptions to be more precise, helps us puny humans that don't have context to understand better and do better review work.

**Comment by acud** [comment]
@nolash the issue is as follows:
In the old syncer an upstream peer would tell the downstream peer on which bins to ask for chunks on (subscriptions). this would guarantee, for example, that the upstream peer would establish subscriptions on bins within depth, although that upstream peer might have the downstream peer outside of it's saturation depth. this would also always guarantee that all chunks will get synced from the downstream peer at some point or another (since subscriptions were established according to the downstream peer's perspective).

in the new stream this is not the case. upstream peers establish streams according to their own kademlia and regardless of the downstream peer's connectivity. while this is more resilient implementation (and less attack prone), it did not in its naive implementation cater for the fact that the downstream peer might be the closest node to certain chunks. the connectivity pattern that was manifesting was that the downstream peer appeared outside of saturation depth of other nodes, thus syncing of certain chunks was not bound to happen (since the downstream peer was the closest one and the upstream peers all had not been subscribed to that node's higher bins since it was outside of their depth). this therefore manifested in a localstore leak since some chunks were never `Set` therefore never added to `gcIndex` therefore never deleted.

File: storage/localstore/mode_put.go [hunk 5]
```diff
@@ -213,47 +197,120 @@ func (db *DB) putRequest(batch *leveldb.Batch, binIDs map[uint8]uint64, item she
...
+			return false, 0, err
+		}
+	}
+
+	return false, gcSizeChange, nil
+}
+
+// setGC is a helper function used to add chunks to the retrieval access
+// index and the gc index in the cases that the putToGCCheck condition
+// vets a gc set. this is to mitigate index leakage in edge cases where
```
[the remark this answers is not in the document]
Comment (acud) [reply storage/localstore/mode_put.go]: warrants

File: storage/localstore/mode_put.go [hunk 6]
```diff
@@ -213,47 +197,120 @@ func (db *DB) putRequest(batch *leveldb.Batch, binIDs map[uint8]uint64, item she
...
+// a chunk is added to a node's localstore and given that the chunk is
+// already within that node's NN (thus, it can be added to the gc index
+// safely)
+func (db *DB) setGC(batch *leveldb.Batch, item shed.Item) (gcSizeChange int64, err error) {
+	if item.BinID <= 0 {
```
[the remark this answers is not in the document]
Comment (janos) [reply storage/localstore/mode_put.go]: Fixed.

**Comment by nolash** [comment]
@janos I think I understand more or less. Thanks.

**Comment by nolash** [comment]
Still not sure about the "what is anonmous content syncing on light nodes" though :)

**Pull request merged**